Kpi sitewide

azizullahsaeidi opened the pull request:
1. Get top 10 viewed resources
2. Get total views
3. Get total registered users views
4. Get total guest users views


Please run below commands
```
php artisan migrate --path=/database/migrations/2024_05_01_124155_add_new_index_to_resource_views_table.php
```

```
php artisan migrate --path=/database/migrations/2024_05_02_181139_add_index_to_resources_table.php
```

```
php artisan migrate --path=/database/migrations/2024_07_01_060829_create_devices_table.php
```
```
php artisan migrate --path=/database/migrations/2024_07_01_062057_create_platforms_table.php
```

```
php artisan migrate --path=/database/migrations/2024_07_01_062632_create_browsers_table.php
```

```
php artisan migrate --path=/database/migrations/2024_07_01_065252_create_sitewide_page_views_table.php
```

```
php artisan migrate --path=/database/migrations/2024_07_03_154638_create_glossary_page_views_table.php
```


```
composer install
```

azizullahsaeidi commented:
> Hi @azizullahsaeidi, thank you for working on this feature. I have a couple of questions before starting to review this:
> 
> 1. Why do we list 'top 10 resources' under the sitewide route? Shouldn't it be 'top 10 pages' instead?
> 2. What are we doing to eliminate bot views?
> 
> Thank you!

Thank you Aaron, yes you are right it should be top 10 pages.

uyscuti-wiki commented:
Thank you @azizullahsaeidi for working on this. Few comments:

1. Are all the filters applicable for all the cards? If not, they should have indication as to which ones are applicable for which.
2. Please add some space between the two rows of the filter. They are too close. 
![image](https://github.com/ddlibrary/ddlibrary/assets/13994249/20c27bf2-00c6-4f86-87a0-5b01e869e185)
3. 'Page Type' should be 'Page type'. 'Is Rot' should be 'Is robot?'.
4. Let's rename all instances of 'sitewides' to its singular form - 'sitewide'. This includes the name of the migration as well. 
5. 'Total view by languages' should be 'Total views by languages'.
6. For 'top 10 viewed pages', let's hyperlink the URL with the URL itself. For example,  [https://library.darakhtdanesh.org/fa/resources](https://library.darakhtdanesh.org/fa/resources) instead of [Resource List](https://library.darakhtdanesh.org/fa/resources). 'Total Top 10 Viewed Pages' should be 'Total views'.
7. For 'total views based on user types', 'Total View' should be 'Total views', and not part of the numbered list, but at the end. 'Unregistered Users Views' should be 'Unregistered', and 'Registered Users Views' should be 'Registered'.
8. 'View based on platform' should be 'Views by platforms'.
9. 'View based on browser' should be 'Views by browsers'.
10. Let's drop the 'view based on device' card altogether.
11. Let's also drop the 'view based on page type' card.

New feature request:
1. Please add a new filter where, when an admin enters the URL of a page, a new card with the page views of that page is returned. ([DDL-163](https://ddlibrary.atlassian.net/browse/DDL-163))

azizullahsaeidi commented:
> Thank you @azizullahsaeidi for working on this. Few comments:
> 
> 1. Are all the filters applicable for all the cards? If not, they should have indication as to which ones are applicable for which.
> 2. Please add some space between the two rows of the filter. They are too close.
>    ![image](https://private-user-images.githubusercontent.com/13994249/347267616-20c27bf2-00c6-4f86-87a0-5b01e869e185.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjA1OTIzNDUsIm5iZiI6MTcyMDU5MjA0NSwicGF0aCI6Ii8xMzk5NDI0OS8zNDcyNjc2MTYtMjBjMjdiZjItMDBjNi00Zjg2LTg3YTAtNWIwMWU4NjllMTg1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MTAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzEwVDA2MTQwNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTZjNzY2ODYyM2JkZTRiZTMyM2M1N2E1Y2Y0NDg5ZmFlYzc5NjdlNmRjMTI0YTliYWI1OGM0NTJkYzk0OGQ0YTImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.3OfmZ4stmh6BQB9nnWF33-Q1RUE6pZ1YFsZyoIdBtDE)
> 3. 'Page Type' should be 'Page type'. 'Is Rot' should be 'Is robot?'.
> 4. Let's rename all instances of 'sitewides' to its singular form - 'sitewide'. This includes the name of the migration as well.
> 5. 'Total view by languages' should be 'Total views by languages'.
> 6. For 'top 10 viewed pages', let's hyperlink the URL with the URL itself. For example,  https://library.darakhtdanesh.org/fa/resources instead of [Resource List](https://library.darakhtdanesh.org/fa/resources). 'Total Top 10 Viewed Pages' should be 'Total views'.
> 7. For 'total views based on user types', 'Total View' should be 'Total views', and not part of the numbered list, but at the end. 'Unregistered Users Views' should be 'Unregistered', and 'Registered Users Views' should be 'Registered'.
> 8. 'View based on platform' should be 'Views by platforms'.
> 9. 'View based on browser' should be 'Views by browsers'.
> 10. Let's drop the 'view based on device' card altogether.
> 11. Let's also drop the 'view based on page type' card.
> 
> New feature request:
> 
> 1. Please add a new filter where, when an admin enters the URL of a page, a new card with the page views of that page is returned. ([DDL-163](https://ddlibrary.atlassian.net/browse/DDL-163))

Thank you Aaron

1. All the filters applicable for all the cards
2. Added more spaces
3. Changed 'Page Type' to 'Page type'. 'Is Rot' to 'Is bot'
4. Changed sitewides to sitewide 
5. 'Total view by languages' changed to 'Total views by languages'. 
6. Now displaying the link instead of the text
7. Changed `Total View` to `Total views`
8. Done
9. Done
10. Done
11. Done

uyscuti-wiki commented on database/migrations/2024_07_01_060809_create_page_types_table.php:
Can we drop this?

azizullahsaeidi replied:
Removed

uyscuti-wiki commented on database/migrations/2024_07_01_065252_create_sitewide_page_views_table.php:
```diff
@@ -0,0 +1,41 @@
+            $table->id();
+            $table->string('page_url', 255);
+            $table->string('user_agent', 255);
+            $table->string('ip_address', 45);
```
Drop this if possible.

azizullahsaeidi replied:
ip_address removed

uyscuti-wiki commented on database/migrations/2024_07_03_154638_create_glossary_page_views_table.php:
```diff
@@ -0,0 +1,44 @@
+        Schema::create('glossary_page_views', function (Blueprint $table) {
+            $table->id();
+            $table->string('user_agent', 255);
+            $table->string('ip_address', 45);
```
Remove IP if possible.

azizullahsaeidi replied:
Removed

uyscuti-wiki commented on resources/views/admin/layouts/navigation.blade.php:
```diff
@@ -93,11 +93,13 @@
         <div class="collapse-divider"></div>
         <h6 class="collapse-header">Analytics</h6>
+        <a class="collapse-item" href="{{ URL::to('admin/analytics/sitewide') }}">Sitewide Analytics</a>
         <a class="collapse-item" href="{{ URL::to('admin/analytics/resources') }}">Resource Analytics</a>
+        <a class="collapse-item" href="{{ URL::to('admin/analytics/glossaries') }}">Glossary Analytics</a>
         <a class="collapse-item" href="{{ URL::to('admin/analytics/users') }}">User Analytics</a>
```
Let's make all routes singular - 'resource', 'glossary' and 'user'.

azizullahsaeidi replied:
Done

uyscuti-wiki reviewed (changes requested):
Thank you Aziz for making the changes. Here are few more minor comments before we get this merged.

azizullahsaeidi commented on app/Traits/SitewidePageViewTrait.php:
```diff
@@ -0,0 +1,73 @@
+                'title' => $title,
+                'page_url' => $request->url(),
+                'user_agent' => $request->userAgent(),
+                'ip_address' => $request->ip(),
```
if you think IP address is not necessary so we can remove it.

azizullahsaeidi commented on app/Models/PageType.php:
I think we should keep page type. by page type we can know the number of visits of news, resources, contact-us, impact.

azizullahsaeidi commented on app/Http/Controllers/GlossaryAnalyticsController.php:
```diff
@@ -0,0 +1,117 @@
+            return $this->filterPageViews($query, $request);
+        })
+            ->groupBy('language')
+            ->select('language', DB::raw('count(*) as count'))
+            ->orderBy('count', 'desc')
```
Done

azizullahsaeidi commented on app/Traits/SitewidePageViewTrait.php:
```diff
@@ -0,0 +1,73 @@
+            $device = Device::firstOrCreate(['name' => $agent->device()]);
+            $pageTypeModel = PageType::firstOrCreate(['name' => $pageType]);
+            $platform = Platform::firstOrCreate(['name' => $agent->platform()]);
+            $browser = Browser::firstOrCreate(['name' => $agent->browser    ()]);
```
Done

azizullahsaeidi commented on app/Traits/SitewidePageViewTrait.php:
```diff
@@ -0,0 +1,73 @@
+
+        return 'page';
+    }
+
```
Removed

azizullahsaeidi commented on app/Traits/GlossaryPageViewTrait.php:
```diff
@@ -0,0 +1,54 @@
+            GlossaryPageView::insert([
+                'title' => $title,
+                'user_agent' => $request->userAgent(),
+                'ip_address' => $request->ip(),
```
Removed

uyscuti-wiki reviewed (approved):
Looks great. Thank you Aziz!

azizullahsaeidi commented on app/Http/Controllers/SitewideAnalyticsController.php:
```diff
@@ -0,0 +1,124 @@
+            ->get();
+
+        $totalGuestViews = $this->getTotalViews($request, 'yes');
+        $totalViewBasedOnLanguage = $this->getTotalViewBasedOnLanguage($request);
```
done

azizullahsaeidi commented on app/Http/Controllers/SitewideAnalyticsController.php:
```diff
@@ -0,0 +1,124 @@
+            return $this->filterPageViews($query, $request);
+        })
+            ->groupBy('language')
+            ->select('language', DB::raw('count(*) as count'))
```
Done
